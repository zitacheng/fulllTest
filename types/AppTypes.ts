export type UserProps = {
    login: string,
    id: number,
    avatar_url: string,
    checked: boolean
  }