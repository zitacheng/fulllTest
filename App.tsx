import React, {useState} from 'react';
import {
  SafeAreaView,
  StatusBar,
  StyleSheet,
  Text,
  useColorScheme,
  View,
  Image,
  Alert,
  TouchableOpacity,
  VirtualizedList,
} from 'react-native';

import {Card} from './components/Card';
import {InputUser} from './components/InputUser';
import {Header} from './components/Header';
import {Menu} from './components/Menu';
import { UserProps } from './types/AppTypes';

// interface CustomResponseHeaders extends Headers {
//   map: {
//     'x-ratelimit-reset'?: string;
//     'x-ratelimit-remaining'?: string;
//     'x-ratelimit-used'?: string;
//   }
// }

// interface CustomResponse extends Response {
//   headers: CustomResponseHeaders;
// }

function App(): React.JSX.Element {
  const isDarkMode = useColorScheme() === 'dark';
  const [selectAll, setSelectAll] = useState<boolean>(false);
  const [edit, setEdit] = useState<boolean>(false);
  const [users, setUsers] = useState<UserProps[]>([]);

  //check the checkbox on card
  const setUserChecked = (user: UserProps) => {
    let idx = users?.findIndex((el) => el.id === user.id);
    
      if (idx != undefined  && idx >= 0 && users) {
        const newArr: UserProps[] = [...users];
        //Uncheck Select All
        if (user.checked == true && selectAll == true) 
        {
          setSelectAll(false);
        }
        if (newArr) {
          newArr[idx].checked = !user.checked;
          setUsers(newArr);
        }
    }
  }

  const getUser = (user: string) => {
    console.log("user ", user)
      fetch('https://api.github.com/search/users?q=' + user)
      .then((response) => {
        console.log("response", response.status)
        // console.log("response", response)
        // console.log("MAP", response.headers.map['x-ratelimit-reset'])
        // console.log("MAP", response.headers.map['x-ratelimit-remaining'])
        // console.log("MAP", response.headers.map['x-ratelimit-used'])
        if (response.status == 200)
          return response; 
        else
          throw response.status;
          // return response; 
      })
      .then(response => response.json())
      .then(json => {
        setUsers(json?.items);
        
        if (json.items?.length == 0)
          Alert.alert('Not found', 'No user has this username.');
      })
      .catch(error => {
        Alert.alert('Error on the request', 'You have exceed the rate limit, please try later');
        console.error(error);
      });
  };

  const getItem = (_data: unknown, index: number): UserProps => {
    return users[index];
  };

  const getItemCount = (_data: unknown) => users.length;

  return (
    <SafeAreaView style={styles.container}>
      <StatusBar barStyle={isDarkMode ? 'light-content' : 'dark-content'} />
      <Header title={'Github Search'} />
      <InputUser setUsers={setUsers} getUser={getUser} />
      <Menu
        edit={edit}
        setEdit={setEdit}
        users={users}
        setUsers={setUsers}
        selectAll={selectAll}
        setSelectAll={setSelectAll}
      />
      <VirtualizedList
        initialNumToRender={2}
        renderItem={({item}) => (
          <Card user={item} edit={edit} setUserChecked={setUserChecked} />
        )}
        keyExtractor={item => item.id.toString()}
        getItemCount={getItemCount}
        getItem={getItem}
      />
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  scroll: {
    padding: 16,
    flex: 1,
  },
});

export default App;
