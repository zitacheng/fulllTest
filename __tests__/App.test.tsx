import * as React from 'react';
import { render, screen, userEvent, fireEvent, waitFor } from '@testing-library/react-native';
import {expect, jest, test} from '@jest/globals';
import {UserProps} from '../types/AppTypes'
import App from '../App';

jest.useFakeTimers();

/**
 * A good place to start is having a tests that your component renders correctly.
 */

test('check snapshot', () => {
  const tree = render(<App />).toJSON();
  expect(tree).toMatchSnapshot();
})

test('renders all elements correctly', () => {
  render(<App />);

  expect(screen.getByRole('header', { name: 'Github Search' })).toBeOnTheScreen();
  expect(screen.getByPlaceholderText("Search Input")).toBeOnTheScreen();
});

test('testing input with 1 result', async () => {
  render(<App />);

  // @ts-ignore
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ items: [{id: 1, avatar_url: '', login: '', checked: false}] }),
    })
  );

  fireEvent(screen.getByPlaceholderText('Search Input'), 'onChangeText', 'zita');
  await waitFor(() => expect(screen.queryAllByRole('none', {name: 'userCard'}).length).toBe(1))
});

test('testing input with 10 result', async () => {
  render(<App />);

  const items : any = []
  for (let i = 0; i < 10; i++) {
    items.push({id: i, avatar_url: '', login: '', checked: false})
  }

  // @ts-ignore
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ items: items }),
    })
  );

  fireEvent(screen.getByPlaceholderText('Search Input'), 'onChangeText', 'zita');
  await waitFor(() => expect(screen.queryAllByRole('none', {name: 'userCard'}).length).toBe(2))
});

test('testing input with 1 result', async () => {
  render(<App />);

  const items : UserProps[] = []
  for (let i = 0; i < 1; i++) {
    items.push({id: i, avatar_url: '', login: '', checked: false})
  }

  // @ts-ignore
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ items: items }),
    })
  );

  fireEvent(screen.getByPlaceholderText('Search Input'), 'onChangeText', 'zita');
  await waitFor(() => expect(screen.queryAllByRole('none', {name: 'userCard'}).length).toBe(1))
});

test('test pressing edit and checked button', async () => {
  render(<App />);

  const items : UserProps[] = []
  for (let i = 0; i < 5; i++) {
    items.push({id: i, avatar_url: '', login: '', checked: false})
  }

  // @ts-ignore
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ items: items }),
    })
  );

  fireEvent(screen.getByPlaceholderText('Search Input'), 'onChangeText', 'zita');
  await waitFor(() => expect(screen.queryAllByRole('none', {name: 'userCard'}).length).toBe(2))

  expect(screen.queryAllByRole('image', {name: 'checkedImg'}).length).toBe(0)

  fireEvent.press(screen.getByRole('button', {name: 'editBtn'}));
  fireEvent.press(screen.getByRole('button', {name: 'selectAllButton'}));

  await waitFor(() => expect(screen.queryAllByRole('image', {name: 'checkedImg'}).length).toBe(2))
});


test('test deleting users', async () => {
  render(<App />);

  const items : UserProps[] = []
  for (let i = 0; i < 5; i++) {
    items.push({id: i, avatar_url: '', login: '', checked: false})
  }

  // @ts-ignore
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ items: items }),
    })
  );

  fireEvent(screen.getByPlaceholderText('Search Input'), 'onChangeText', 'zita');
  await waitFor(() => expect(screen.queryAllByRole('none', {name: 'userCard'}).length).toBe(2))

  fireEvent.press(screen.getByRole('button', {name: 'editBtn'}));
  fireEvent.press(screen.getByRole('button', {name: 'selectAllButton'}));
  fireEvent.press(screen.getByRole('button', {name: 'deleteButton'}));

  await waitFor(() => expect(screen.queryAllByRole('none', {name: 'userCard'}).length).toBe(0))
  });
