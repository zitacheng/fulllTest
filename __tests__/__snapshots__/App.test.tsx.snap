// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignContent": "center",
        "alignSelf": "center",
        "backgroundColor": "#929292",
        "flexDirection": "row",
        "height": 64,
      }
    }
  >
    <Text
      accessibilityRole="header"
      style={
        {
          "alignSelf": "center",
          "fontSize": 20,
          "fontWeight": "500",
          "textAlign": "center",
          "width": "100%",
        }
      }
    >
      Github Search
    </Text>
  </View>
  <TextInput
    autoCapitalize="none"
    autoCorrect={false}
    onChangeText={[Function]}
    placeholder="Search Input"
    style={
      {
        "backgroundColor": "#d5d5d5",
        "borderRadius": 6,
        "fontSize": 20,
        "marginBottom": 24,
        "marginLeft": 16,
        "marginRight": 16,
        "marginTop": 24,
        "padding": 10,
      }
    }
    value=""
  />
  <View
    accessibilityLabel="editBtn"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignSelf": "flex-start",
        "backgroundColor": "#04a1fe",
        "borderRadius": 18,
        "marginBottom": 5,
        "marginLeft": 16,
        "opacity": 1,
        "padding": 8,
      }
    }
  >
    <Text>
      Edit
    </Text>
  </View>
  <RCTScrollView
    getItem={[Function]}
    getItemCount={[Function]}
    initialNumToRender={2}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
  >
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;
